### ISSUE TITLE
MEGA-ISSUE: Syntax highlighting in `language-gfm` (Markdown)

### ISSUE THREAD

**Issue by savetheclocktower** [pulsar-edit/pulsar#878]
### IMPORTANT: Some issues have already been fixed!

If you’re still on the regular 1.113 release, you might be suffering from a problem that has already been fixed. Many fixes landed on `master` in #859. You are encouraged to download the latest [rolling release](https://pulsar-edit.dev/download.html#rolling-release) — both (a) to see whether what you’re reporting has been fixed, and (b) so that you can enjoy the fixes that have been reported since the 1.113 release.

---

This will serve as a catch-all issue for any problems you may encounter with syntax highlighting in Markdown files (`language-gfm`). If you have problems with the new syntax highlighting (or folds, or indents) that are specific to `md` or `markdown` files, keep reading.

### Something isn't highlighting correctly!

If you’ve found any highlighting regressions since 1.113:

First, please **scroll down and see if someone else has reported the issue**. If so, then you need only sit back and wait for a fix — most issues will be fixed in version 1.114!

If not, please comment with

* A small amount of sample code to reproduce the issue
* Screenshots (before and after would be ideal, but just the “after” is OK for obvious things

### I want to go back to the old highlighting!

You can easily opt out of the new Tree-sitter highlighting for any language, but first please:

* **subscribe to this issue** so you'll know when the problem is fixed
* **remove the config change** when the next release comes out so that you can enjoy the new grammar once again

To revert to the TextMate-style grammar for this language only, add the following to your `config.cson`:

```coffeescript
".gfm.source":
  core:
    useTreeSitterParsers: false
```

**Comment by savetheclocktower** [comment]
@Atelier-Mirai [has reported](https://github.com/pulsar-edit/pulsar/issues/867) an occasional failure in Markdown parsing when typing Japanese text.

This issue is broadly understood, but I'd love to get some specific reproduction steps, particularly if it's something that I can cut and paste into a document without having to understand the particulars of Japanese keyboard input. If I can make the parser fail consistently in the manner described in #867, I can explore various solutions for recovering from the error.

**Comment by savetheclocktower** [comment]
_I meant to post this explanation in here, but inadvertently put it in #867. Let me try again._

So I'll try to summarize the problem for anyone who might be reading. **If @Atelier-Mirai's description gives you ideas on how to reproduce this, please share them.**

[The `tree-sitter-markdown` parser we use](https://github.com/ikatyang/tree-sitter-markdown) relies heavily on an external C scanner. [Emscripten](https://emscripten.org/) is the tool that turns Tree-sitter parsers into `.wasm` files, and it typically knows how to deal with scanners that rely on functions defined in C.

But `tree-sitter-markdown` [makes heavy use of `assert`](https://github.com/ikatyang/tree-sitter-markdown/blob/master/src/scanner.cc#L56) — which isn't a standard library function as such, but rather a macro that performs basic sanity checks. When you write `assert(foo)`, you're generally saying that if `foo` is true, then something very strange has happened, and the program should exit. I've been researching it, but I can't find evidence of Emscripten being able to handle `assert`. (The `ASSERTIONS` flag that the message describes appears to be a separate thing.)

The parser [has an “avoid crash” mode](https://github.com/ikatyang/tree-sitter-markdown/blob/master/src/scanner.cc#L8-L12) where, instead of using `assert`s, it throws exceptions that are then caught at the top-level scan function, which then tries to recover despite the violated assumptions. Except that Emscripten doesn't support exceptions by default; you can enable it, but it is described as having “[relatively high overhead](https://emscripten.org/docs/porting/exceptions.html#emscripten-javascript-based-exception-support).” It sounds like a thrown exception has to cross the bridge from WASM back out into JavaScript. I'd love to avoid that if I could. That page [also mentions](https://emscripten.org/docs/porting/exceptions.html#webassembly-exception-handling-based-support) a native WebAssembly exceptions proposal that is probably implemented in Chrome by now, and could probably be an option once we upgrade Electron, but not on the version we're on right now.

It would help if I could reproduce an example of the parser failing one of these sanity checks, but ultimately I've got to figure out how the parser should behave in these scenarios if we can't use `assert` _or_ C++ exceptions. It's frustrating that lots of Tree-sitter parsers — including some first-party parsers! — don't seem to routinely test themselves in the `web-tree-sitter` bindings.

There is [another `tree-sitter-markdown` parser](https://github.com/MDeiml/tree-sitter-markdown) that is actively maintained, and I'd love to use it, but there are some show-stopping bugs I can't seem to get around. I spent several hours last night revisiting this parser to see if any of them were fixed, but [this issue I filed last March](https://github.com/MDeiml/tree-sitter-markdown/issues/92) is still present in Pulsar. (It's always possible that this is an intractable `web-tree-sitter` bug, or that I'm just making a simple error, but [this issue filed in `nvim-treesitter` is reassuring](https://github.com/nvim-treesitter/nvim-treesitter/issues/4716) because it means this is probably a subtle parser bug.)

Ultimately the answer might be to fork one of these parsers — I hate it, but I have to at least consider the idea — and change how it behaves. Tree-sitter was meant to be used for editors, and a parser that fails the first time its assumptions are broken is not something that works in an editor environment where a document may expect to vacillate between validity and invalidity whenever the user types.

**Comment by savetheclocktower** [comment]
@luixxiul noted in #885 that `tree-sitter-markdown` fails to detect URLs in certain circumstances, and I'm not surprised that multibyte input seems to be present in this issue as well. Meanwhile, we can use `tree-sitter-hyperlink` and inject into any paragraphs (or paragraph fragments) to catch the URLs that `tree-sitter-markdown` misses.

**Comment by savetheclocktower** [comment]
OK, that was pretty easy. [This commit](https://github.com/pulsar-edit/pulsar/pull/859/commits/e1e12ff5bace341fff5178157434e368283efb7f) adds auto-linking for URLs that are missed by the Markdown parser. Please read #859 if you want a build in which this issue is fixed without having to wait for 1.114.

**Comment by savetheclocktower** [comment]
@liuxiong332 also noticed that a horizontal rule (usually three or more asterisks, hyphens, or underscores on their own line) was no longer being syntax highlighted. This happened because the original scope for a horizontal rule was `comment.hr.gfm`, and a horizontal rule is just not a comment by any definition of the term.

There isn't any prior art that I can find here, but I decided it should live in the `markup` scope namespace and created `markup.horizontal-rule` instead. I [added styles](https://github.com/pulsar-edit/pulsar/pull/859/commits/ee220dd6efd578c4e42374e076221c0c07997698) for `markup.horizontal-rule` to each of the eight built-in syntax themes; for continuity's sake, I made their styles match those of a comment in each theme.

If you use a community theme that is actively maintained, please ask them to support `markup.horizontal-rule`. If your community theme is not maintained, ask for help in this ticket and we can give you some CSS to add to your user stylesheet to achieve the same effect.

**Comment by Daeraxa** [comment]
[orionaegis7](https://www.reddit.com/user/orionaegis7) on Reddit found the following issue: https://www.reddit.com/r/pulsaredit/comments/199d4bx/unable_to_collapse_tabbed_lines_in_pulsar/

Take the following Markdown text:
```md
Line 1

    Indented Line 1
    Indented Line 2

Line 2
```

In the TextMate grammar a fold icon is provided on `Line 1` to collapse the indented lines.
![image](https://github.com/pulsar-edit/pulsar/assets/58074586/198764b5-1f2d-4b17-b826-020f879fca9e)
![image](https://github.com/pulsar-edit/pulsar/assets/58074586/9896e4f3-201f-470c-8ff9-7043df02f3a5)

This is not present using the Tree-sitter grammar.

**Comment by savetheclocktower** [comment]
This is a limitation of how folds work in Tree-sitter versus TextMate.

Now they're completely semantic instead of based on indentation — good when you want to fold things whether or not they're “indented properly,” bad if you want to fold things in a way that doesn't correspond to semantics.

This is on my radar, though. I don't have it fleshed out, but the fact that folds can now come from

* Tree-sitter
* [Language servers](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange)
* A naïve analysis of the leading indentation level of each individual line of a buffer

means that this is a place where, just like `symbols-view`, we can pick a default and allow community packages to suggest alternatives.

**Comment by bobics** [comment]
Triple backtick code highlighting (green) stopped working:
<img width="229" alt="Screenshot 2024-01-26 at 11 28 04 AM" src="https://github.com/pulsar-edit/pulsar/assets/33119/e3bbecca-35f6-472a-83cb-2be4da7594eb">

**Comment by savetheclocktower** [comment]
> Triple backtick code highlighting (green) stopped working: <img alt="Screenshot 2024-01-26 at 11 28 04 AM" width="229" src="https://private-user-images.githubusercontent.com/33119/300091259-e3bbecca-35f6-472a-83cb-2be4da7594eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYzMDczODksIm5iZiI6MTcwNjMwNzA4OSwicGF0aCI6Ii8zMzExOS8zMDAwOTEyNTktZTNiYmVjY2EtMzVmNi00NzJhLTgzY2ItMmJlNGRhNzU5NGViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDIyMTEyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1OWQ4YTk0NzZjMWVhMDZjYmY1NjUyODlkY2U4MGM5ZmRlOGIwZWQxZDYzNTc1MmE4NWU5MzlhNGE1MWNiY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UbuiXjVCFzywsDxBBlWKNP0_vwfDiuLyYCiMx2rHF0k">

Good catch; it's missing `markup.raw.block.gfm`. I'll add that soon.

**Comment by Daeraxa** [comment]
I've spotted some more differences which I think are gfm specific (or at least GitHub's implementation of it).

## Issue and mention tags:

Tree-sitter:
![image](https://github.com/pulsar-edit/pulsar/assets/58074586/447aaa83-ff81-46bf-92f9-efa051808819)

Textmate:
![image](https://github.com/pulsar-edit/pulsar/assets/58074586/1441ff71-c387-4f4c-9503-afb4998bb197)

The number (I guess this is designed for highlighting issues + PRs) applies the following:
`syntax--variable syntax--issue syntax--tag syntax--gfm`
`syntax--string syntax--issue syntax--number syntax--gfm`

The mention applies:
`syntax--variable syntax--mention syntax--gfm`
`syntax--string syntax--username syntax--gfm`

Not sure how this should with the conflict that is reference links though. e.g. if you have:
```
[@name]
...
[@name]: https://name.com/
```

Should `[@name]` show as a link or as a mention?

**Comment by savetheclocktower** [comment]
This is a known issue. The `tree-sitter-markdown` parser doesn't do anything special with regard to constructs that are only meaningful in a GitHub context.

We could address this by writing our own Tree-sitter parser to detect things like `#123` and `@handler` and injecting it into paragraphs, much like I did with the `language-hyperlink` and `language-todo`. It's just a matter of whether we think it's worth the effort.

By the way, @Daeraxa: instead of digging into the source code next time, run the **Editor: Log Cursor Scope** command and you'll get the uncomplicated versions of scope names in a notification, listed from most general to most specific.

**Comment by Daeraxa** [comment]
Thats a much better way of doing it, I was only grabbing it out of the dev tools though so it wasn't much extra effort in that sense.

I suspected this was the case already I just wanted to get it added to the mega issue so at least we can make comment on it ahead of time.

It is also a bit of a weird one as obviously mentions and issue numbers are not part of markdown (nor part of the GFM spec, these are entirely for GitHub the site and honestly it is kind of odd to format them at all if actually just following a spec as these are not part of it, AFAIK it is only implemented as part of cmark-gfm and linguist.

**Comment by savetheclocktower** [comment]
Hell, I've never loved that the bundle is called `language-gfm` and that the root scope is `source.gfm` instead of something like `text.markdown` (or even `text.html.markdown`), but it's just one of the many scope-related sins that the TextMate-style grammar commits.

**Comment by savetheclocktower** [comment]
@jqueraltmo, @Atelier-Mirai, and anyone else who's interested:

We've been able to get a bug fixed (we hope!) in `MDeiml/tree-sitter-markdown` that was preventing us from being able to use it in Pulsar. This is an actively maintained Markdown Tree-sitter parser, unlike `ikatyang/tree-sitter-markdown`.

I was never able to reproduce the specific assertion failures you folks have reported, so I have no idea if this parser suffers from the same issues; but I'm hopeful it doesn't.

I've made [a language bundle with a grammar based on the new parser](https://web.pulsar-edit.dev/packages/language-markdown-alpha), so please try it out and let me know if it's an improvement. Be sure to **follow the instructions in the `README` file** — in particular, you'll need to **temporarily disable** the builtin `language-gfm` package.

If it truly is an improvement, my hope is that we'll be able to ship it in 1.114, or 1.115 at the latest.

**Comment by jqueraltmo** [comment]
@savetheclocktower, just installed this. I reactivated bracket matching, and it's working fine for now. I'll let you know if there are any crashes.

**Comment by armfem** [comment]
I made the installation Monday, I have had no crash since then. I'll come back if anything changes in the future.

In any case, thanks for the upgrade. It is great to modify markdown files without having to reinitialize Pulsar :smiley:

**Comment by savetheclocktower** [comment]
The “experimental” replacement parser for Markdown is **now the default** as of the just-released 1.116.

After downloading and installing 1.116, you may uninstall `language-markdown-alpha` and re-enable the builtin `language-gfm` package.

Thanks to everyone who helped me test it. It's got a few minor issues around the handling of italics, but ultimately that's much better than the situation we had with the previous parser. I'll probably keep this thread open for another month or so just to see if there are any new issues reported.

**Comment by brunobord** [comment]
Hello. A very happy Atom-then-Pulsar user here, but apparently, it seems that the syntax highlighting issue doesn't seem to be solved for SQL code blocks in Markdown files.

I've made the test today with the most recent release (1.119) and it's still there.

I've ran all the following examples with `pulsar --safe`, using the Atom Dark UI and themes (they're not my personal favorite, but I wanted to make sure that the syntax highlight issue was not because of a custom preference).

Here's the rendering of a `test.sql` file:
![test.sql rendering screen capture](https://github.com/user-attachments/assets/f276c15a-3ff4-4e8b-bbe3-e5a9d09477c0)


Now here is the same code, in a test.md file.

![test.md rendering screen capture](https://github.com/user-attachments/assets/988315a6-3d02-41e5-881b-9053300a5b70)

As you can see at the bottom of the second screen capture, the python code block is correctly highlighted, so I can safely say that, at least for SQL code blocks, the feature doesn't work properly.

Could you please investigate on that?

**Comment by savetheclocktower** [comment]
> As you can see at the bottom of the second screen capture, the python code block is correctly highlighted, so I can safely say that, at least for SQL code blocks, the feature doesn't work properly.

This is a known issue. Injections don't work across different grammar types. Most built-in languages now have modern Tree-sitter grammars, but SQL doesn't.

The way for Pulsar to fix this is to find a good `tree-sitter-sql` implementation (I see a few candidates just through searching, but I haven't tried them out yet) and create a modern Tree-sitter grammar for SQL.

The workaround for you would be to uncheck “Hide Duplicate Grammars” in the `grammar-selector` settings and opt into the older “GitHub Markdown” grammar via the grammar selector menu. The Markdown highlighting won't be as accurate, but it'd at least return you to the status quo you remember from Atom. If you wanted to make it permanent, you could follow the instructions in the issue description (under “I want to go back to the old highlighting!”) — or you could just switch to the older Markdown grammar as needed when you want SQL highlighting in your Markdown files.

**Comment by brunobord** [comment]
> This is a known issue. Injections don't work across different grammar types. Most built-in languages now have modern Tree-sitter grammars, but SQL doesn't.
> The way for Pulsar to fix this is to find a good tree-sitter-sql implementation (I see a few candidates just through searching, but I haven't tried them out yet) and create a modern Tree-sitter grammar for SQL.

Ok, I see now. I also see the options I have.
Best of luck for finding the right tree-sitter. Many thanks.

**Comment by savetheclocktower** [comment]
I've barely heard complaints about Markdown highlighting ever since we switched Tree-sitter parsers, so I'm closing this issue!
